### DISCUSSION TITLE
Wakapi stuck at 20231023-fill_last_branch migration after update to 2.9.3

### DISCUSSION THREAD

**Discussion by Kovah** [muety/wakapi#551] [Q&A] [answered]
Hi,
I've just updated Wakapi from some pre 2.9 version to the latest one and after starting the Docker container, Wakapi ist stuck at `[INFO ] potentially running migration '20231023-fill_last_branch'`.
It's been like that for half an hour. Not sure if it's supposed to take that long, but I have not a million entries in the database, so I guess that there's something wrong with that migration.

Is there a way to debug this or get it running somehow?

I have downgraded to 2.9.2 and it's now running without starting any other migration.

**Comment by muety** [comment]
What database are you using?

**Reply by muety** [reply, marked as answer]
And how many entries do you have in your heartbeats table? As already mentioned in the code, the migration is implemented in a super inefficient way where it iterates through every single heartbeat one by one. So depending on the data volume, it might well take a couple of hours. I can't think of a way where the migration gets "stuck" in idle. To see if the migration is still running, you could check the active database queries like so:

**MySQL**
```sql
show processlist
```

**Postgres**
```
select * from pg_stat_activity;
```

I'd probably just let it run overnight. Sorry for that.

**Reply by muety** [reply]
Alternatively, if you don't care about properly filling the `<<LAST_BRANCH>>` placeholder for historic data (e.g. because you didn't use the browser plugin anyway), you can also just manually add a key-value pair `("20231023-fill_last_branch", "done")` to your `key_string_values` table, which will cause the migration to be skipped from then on.

**Reply by Kovah** [reply]
Thanks for the fast reply. I'm using MySQL and currently have 200.000 entries in the heartbeat table. I guess that even with a really slow operation, it shouldn't take like 20 minutes to run those queries, right?
I'l trigger the update  before going to bed and see what happens during the night.

**Reply by Kovah** [reply]
Started the update again, looks like there's something happening:

```
| Id     | User | Host             | db       | Command | Time | State               | Info                                                                                                 | Progress |
|--------|------|------------------|----------|---------|------|---------------------|------------------------------------------------------------------------------------------------------|----------|
| 373931 | root | 172.20.0.9:55136 | wakatime | Execute |    0 | Creating sort index | SELECT * FROM `heartbeats` WHERE `heartbeats`.`user_id` = ? AND `heartbeats`.`project` = ? AND `bran |    0.000 |
| 373932 | root | 172.20.0.9:55148 | wakatime | Execute |    0 | Creating sort index | SELECT * FROM `heartbeats` WHERE `heartbeats`.`user_id` = ? AND `heartbeats`.`project` = ? AND `bran |    0.000 |
```
Well, hope it's done tomorrow. :D

**Reply by Kovah** [reply]
The update was finished. According to the uptime monitoring, it took about 40 minutes. Maybe it's worth being mentioned in the release notes so others know there will be a longer downtime.
Thanks for the help!
